### PR TITLE
Add support for H264 SVC

### PR DESCRIPTION
**Pull request by prtmD**
Contribution from Ittiam Systems Pvt. Ltd.
[Mediasoup_Contributions_H264_SVC_v1_0.pdf](https://github.com/versatica/mediasoup/files/8350639/Mediasoup_Contributions_H264_SVC_v1_0.pdf)
Author: Pratima Dhuldhule

### CODE REVIEW

**Comment by jmillan** [comment]
I'll check this next week.

File: worker/include/RTC/Codecs/Tools.hpp [hunk 1]
```diff
@@ -70,6 +72,12 @@ namespace RTC
 
 								break;
 							}
+                            case RTC::RtpCodecMimeType::Subtype::H264_SVC:
```
Comment (ibc) [inline worker/include/RTC/Codecs/Tools.hpp]: Wrong indentation here. Please run "make format" in worker folder.

File: worker/src/RTC/Codecs/H264_SVC.cpp [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+					}
+				}
+			}
+
+			return payloadDescriptor.release();
+		}
+
+		std::unique_ptr<H264_SVC::PayloadDescriptor> H264_SVC::ParseSingleNalu(const uint8_t* data, size_t len, std::unique_ptr<H264_SVC::PayloadDescriptor> payloadDescriptor, bool isStartBit)
+		{
+
```
Comment (ibc) [inline worker/src/RTC/Codecs/H264_SVC.cpp]: No blank line here please

File: worker/src/RTC/Codecs/H264_SVC.cpp [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+					byte = data[offset];
+
+					payloadDescriptor->tlIndex = byte >> 5 & 0x03;
+
+					payloadDescriptor->hasSlIndex = payloadDescriptor->slIndex ? true : false;
+					payloadDescriptor->hasTlIndex = payloadDescriptor->tlIndex ? true : false;
+
+					break;
+
```
Comment (ibc) [inline worker/src/RTC/Codecs/H264_SVC.cpp]: No blank line

File: worker/src/RTC/Codecs/H264_SVC.cpp [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+            MS_ASSERT(context->GetTargetSpatialLayer() >= 0, "target spatial layer cannot be -1");
+			MS_ASSERT(context->GetTargetTemporalLayer() >= 0, "target temporal layer cannot be -1");
+
+            auto packetSpatialLayer  = GetSpatialLayer();
+			auto packetTemporalLayer = GetTemporalLayer();
+			auto tmpSpatialLayer     = context->GetCurrentSpatialLayer();
+			auto tmpTemporalLayer    = context->GetCurrentTemporalLayer();
+
+            // If packet spatial or temporal layer is higher than maximum announced
```
Comment (ibc) [inline worker/src/RTC/Codecs/H264_SVC.cpp]: Wrong indentation

File: worker/src/RTC/Codecs/H264_SVC.cpp [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+
+			// Update current spatial layer if needed.
+			if (tmpSpatialLayer != context->GetCurrentSpatialLayer())
+				context->SetCurrentSpatialLayer(tmpSpatialLayer);
+
+			// Update current temporal layer if needed.
+			if (tmpTemporalLayer != context->GetCurrentTemporalLayer())
+				context->SetCurrentTemporalLayer(tmpTemporalLayer);
+
+
```
Comment (ibc) [inline worker/src/RTC/Codecs/H264_SVC.cpp]: Just one blank line here please

File: worker/test/include/helpers.hpp [hunk 6]
```diff
@@ -25,6 +27,125 @@ namespace helpers
...
+	    if(!addToBuffer(buf, &packet_size, &oneByte, 1))
+	        return false;
+
+	    // write payload
+	    if(!addToBuffer(buf, &packet_size, payload, nalLength))
+	        return false;
+
+	    *len = packet_size;
+	    // write the RTP packet to a binary file
+//	    std::ofstream ofp(filePath, std::ios::out | std::ios::binary);
```
Comment (ibc) [inline worker/test/include/helpers.hpp]: Wrong indentation and styling in the whole file. "make format" should fix it.

**Review by ibc — CHANGES_REQUESTED** [review]
Looks amazing although I miss changed in supportedCapabilities.ts and also in the similar Rust file, which should include the new codec.

**Comment by prtmD** [comment]
Thank you for your review comments. We will work on these and commit the updates.

**Comment by ibc** [comment]
`make format` required in `worker/` folder.

**Comment by nazar-pc** [comment]
I have fixed a few issues and applied formatting rules, but there are 2 tests from added by this PR that are failing:
```
-------------------------------------------------------------------------------
Scenario: parse RTP packets with H264 SVC
  parse I0-5.bin
-------------------------------------------------------------------------------
../../../test/src/RTC/TestRtpPacketH264Svc.cpp:138
...............................................................................

../../../test/src/RTC/TestRtpPacketH264Svc.cpp:188: FAILED:
  REQUIRE( payloadDescriptor->isKeyFrame == true )
with expansion:
  false == true

-------------------------------------------------------------------------------
Scenario: parse H264_SVC payload descriptor
  parse payload descriptor for NALU 5
-------------------------------------------------------------------------------
../../../test/src/RTC/Codecs/TestH264_SVC.cpp:94
...............................................................................

../../../test/src/RTC/Codecs/TestH264_SVC.cpp:115: FAILED:
  REQUIRE( payloadDescriptor->isKeyFrame == true )
with expansion:
  false == true

===============================================================================
test cases:   42 |   40 passed | 2 failed
assertions: 1785 | 1783 passed | 2 failed
```

@prtmD can you review my changes fix those tests? Apart from that should be good for merge.

**Comment by prtmD** [comment]
Setting of `isKeyFrame` was missed in the `payloadDescriptor` for NALU 5. The tests should build now.

**Comment by nazar-pc** [comment]
Fails to compile on macOS for some reason, for other platforms seems to work

**Comment by prtmD** [comment]
> Fails to compile on macOS for some reason, for other platforms seems to work

From the compile logs I see that it is failing to find `bits/stdc++.h` (will not be present on macOS). I will remove this import and push the code again.

**Review by nazar-pc — APPROVED** [review]
@ibc please take a look at this again

**Comment by ibc** [comment]
Sorry for delay, handling this PR this week.

**Comment by prtmD** [comment]
> Sorry for delay, handling this PR this week.

Thank you. Please let us know if you need anything from our end.

File: worker/src/RTC/Codecs/H264_SVC.cpp [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+		  const uint8_t* data, size_t len, RTC::RtpPacket::FrameMarking* frameMarking, uint8_t frameMarkingLen)
+		{
+			MS_TRACE();
+
+			if (len < 2)
+				return nullptr;
+
+			std::unique_ptr<PayloadDescriptor> payloadDescriptor(new PayloadDescriptor());
+
+			// TODO: Pratima
```
Comment (jmillan) [inline worker/src/RTC/Codecs/H264_SVC.cpp]: What is this TODO?

Comment (prtmD) [reply worker/src/RTC/Codecs/H264_SVC.cpp]: Sorry that was a handled TODO. I have removed it in the new commit.

**Review by jmillan — APPROVED** [review]

**Comment by haoxinllc** [comment]
Does the mediasoup client support h.264-SVC codec?

**Comment by ibc** [comment]
Questions in the forum please.

**Comment by prtmD** [comment]
> Does the mediasoup client support h.264-SVC codec?

mediasoup-client is codec-agnostic. So it should work without needing any changes.

**Comment by haoxinllc** [comment]
> 
Thanks for the guidance
As far as I know, mediasoup Js/C++ client work depends on libwebrtc library, H.264 SVC codec needs to be compiled in libwebrtc library to work? So how to fully apply or test H.264 SVC in the mediasoup client?

**Pull request merged**